pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,316 @@
+   * info.
+   */
+  private static String getPartitionFunction(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    String partitionFunction = kvOptions.get(kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION));
+    if (partitionFunction != null && !partitionFunction.equals(tablePartitionInfo.getPartitionFunctionName())) {
+      LOGGER.debug("Override implicit table hint for {} with explicit partition function: {}",
+          tableScan, partitionFunction);
+      return partitionFunction;
+    } else {
+      return tablePartitionInfo.getPartitionFunctionName();
+    }
+  }
+
+  /**
+   * Get the partition parallelism from the hint, if any, otherwise returns null.
+   */
+  @Nullable
+  private static Integer getPartitionParallelism(LogicalTableScan tableScan, Map<String, String> kvOptions) {
+    String partitionParallelismStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM);
+    if (partitionParallelismStr != null) {
+      try {
+        int partitionParallelism = Integer.parseInt(partitionParallelismStr);
+        LOGGER.debug("Override implicit table hint for {} with explicit partition parallelism: {}", tableScan,
+            partitionParallelism);
+        return partitionParallelism;
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            "Invalid partition parallelism: " + partitionParallelismStr + " for table: " + tableScan);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Get the partition size from the hint, if any, otherwise use the partition size from the table partition info.
+   */
+  private static int getPartitionSize(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    int partitionSize = tablePartitionInfo.getNumPartitions();
+    String partitionSizeStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+    if (partitionSizeStr != null) {
+      try {
+        int explicitPartitionSize = Integer.parseInt(partitionSizeStr);
+        if (explicitPartitionSize != partitionSize) {
+          LOGGER.debug("Override implicit table hint for {} with explicit partition size: {}", tableScan,
+              explicitPartitionSize);
+          partitionSize = explicitPartitionSize;
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("Invalid partition size: " + partitionSizeStr + " for table: " + tableScan);
+      }
+    }
+    return partitionSize;
+  }
+
+  /**
+   * Get the table partition info for the given table scan, transforming Calcite's table name to Pinot's table name
+   * and adding the table type if needed.
+   */
+  @Nullable
+  private TablePartitionInfo getTablePartitionInfo(LogicalTableScan tableScan) {
+    String tableName = RelToPlanNodeConverter.getTableNameFromTableScan(tableScan);
```

Comment:
The name might already have type suffix. We should try to extract the type first

Reply:
Changed. If we receive an offline or online table name, we just return the TablePartitionInfo associated with it and to not check whether the other is present.